text/0249-v2-experiments.md

```diff
@@ -0,0 +1,380 @@
+clearly marked as such).
+
+# Proposed changes
+
+To achieve the goals of this RFC, we propose the following changes:
+
+## 1. No unstable code in the main mono-CDK modules
+
+Because of a combination of how `aws-cdk-lib` will be depended on by third-party
+libraries (through peer dependencies), and the goals of ths RFC, it will no
+longer be possible to vend unstable code as part of `aws-cdk-lib`'s main modules
+(so, modules that are marked "stable" will no longer have the capability to have
+unstable APIs inside of them).
+
+We will need to add validations to our build scripts that make sure this rule is
+never broken.
+
+### Why can't we vend unstable code in mono-CDK stable modules?
+
+This paragraph explains why it's not possible to vend unstable code in the
+mono-CDK main modules. If you're already convinced that that's true, feel free
+to skip to the next section.
+
+<details>
+Imagine we could ship unstable code in `aws-cdk-lib`.
+The following scenario would then be possible:
+
+Let's say we have a third-party library, `my-library`, that vends `MyConstruct`.
+It's considered stable by its author. However, inside the implementation of
+`MyConstruct`, it uses an experimental construct, `SomeExperiment`, from
+mono-CDK's S3 module. It's just an implementation detail, though; it's not
+reflected in the API of `MyConstruct`.
+
+`my-library` is released in version `2.0.0`, and it has a peer dependency on
+`aws-cdk-lib` version `2.10.0` (with a caret, so `"aws-cdk-lib": "^2.10.0"`).
+
+Some time passes, enough that `aws-cdk-lib` is now in version `2.20.0`. A CDK
+customer wants to use `my-library` together with the newest and shiniest
+`aws-cdk-lib`,`2.20.0`, as they need some recently released features. However,
+incidentally, in version `2.12.0` of `aws-cdk-lib`, `SomeExperiment` was broken
+-- which is fine, it's an experimental API. Suddenly, the combination of
+`my-library` `2.0.0` and `aws-cdk-lib` `2.20.0` will fail for the customer at
+runtime, and there's basically no way for them to unblock themselves other than
+pinning to version `2.11.0` of `aws-cdk-lib`, which was exactly the problem
+mono-CDK was designed to prevent in the first place.
+
+</details>
+
+## 2. Separate unstable code from mono-CDK main modules
+
+As a consequence of the above point, we need to move all unstable code out of
+the mono-CDK main modules.
+
+There a few possible options where the unstable code can be moved. They all have
+their advantages and disadvantages.
+
+**Note**: I purposefully don't mention the issue of how should this decision
+affect the number of Git repositories the CDK project uses. I consider that an
+orthogonal concern to this one, and more of an implementation detail. The number
+of Git repositories is also a two-way door, unlike the package structure
+decision, which I believe cannot be changed without bumping the major version of
+CDK.
+
+### Option 1: sub-package of `aws-cdk-lib`
```

Comment:
This option conflicts with "No unstable code in the main mono-CDK modules". I don't see a reason to propose it.